frappe/public/js/legacy/form.js

```diff
@@ -922,6 +922,16 @@ _f.Frm.prototype.action_perm_type_map = {
 
 _f.Frm.prototype.validate_form_action = function(action) {
 	var perm_to_check = this.action_perm_type_map[action];
+	var allowed_for_workflow = false;
+	var perms = frappe.perm.get_perm(this.doc.doctype)[0];
+
+	// Allow submit, write and create permissions for read only documents that are assigned by 
+	// workflows if the user already have those permissions. This is to allow for users to 
+	// continue through the workflow states and to allow execution of functions like Duplicate.
+	if (frappe.workflow.is_read_only(this.doctype, this.docname) && (perms["write"] ||
+		perms["create"] || perms["submit"])) {
+		var allowed_for_workflow = true;
```

Comment:
When you configure a workflow with "Only Allow Edit for" set for a particular role. The entire form is set as read only. The thing is, set_read_only actually sets the read permission but strips all other permissions. 

So for a document which is subject to a workflow, a user with sufficient permissions would not be able to go to the next step of the workflow because the workflow action actually requires write permissions to store the workflow state. This is also the same for the "Duplicate" function which requires create permissions.